sympy/sets/conditionset.py

```diff
@@ -191,8 +191,12 @@ def __new__(cls, sym, condition, base_set=S.UniversalSet):
 
     @property
     def free_symbols(self):
-        s, c, b = self.args
-        return (c.free_symbols - s.free_symbols) | b.free_symbols
+        cond_syms = self.condition.free_symbols - self.sym.free_symbols
+        return cond_syms | self.base_set.free_symbols
+
+    @property
+    def bound_symbols(self):
```

Comment:
I think this should be removed.  What is the need to identify bound symbols? Nothing in this PR uses that. And knowing what they, how does that provide any useful unformation? `subs` would still have to figure out what and where to treat a given symbol as bound.  

e.g. in the test `ConditionSet(x, x < 10, ImageSet(Lambda(y, y**2), S.Integers)` the symbols are x and y but there are no free symbols so both x and y are bound. The set could have been written with only 1 bound symbol, too (as `as_dummy` of this shows). That was something I realized as I was reworking the `as_dummy` routines to allow better canonical expressions to be obtained.

So I am not convinced we need to start adding `bound_symbols` to the code base: they are just `e.atoms(Symbol) - e.free_symbols`.

Reply:
> So I am not convinced we need to start adding `bound_symbols` to the code base: they are just `e.atoms(Symbol) - e.free_symbols`.

I'm starting to realize that too. And I'm also wondering now if it's at all useful to have `bound_symbols` everywhere. Take for example:
```
e = x + Sum(x, (x, 0, 10))
```
Its `free_symbols` are `x` either way, even if the bound variable is renamed. As for `e.bound_variables`, it'd be highly misleading if the answer were `{x}`. I guess it all depends on what `bound_variables` are used for but I'm not sure they should be anything but `e.atoms(Symbol) - e.free_symbols` as @smichr just said.
Maybe the current approach is not that bad after all: have it as a kind of marker on those special classes that `.as_dummy` should treat in a special way. And actually specifying its value par class instead of just putting `atoms - free_symbols` would then make the implementation clearer.

As for the tests I partly wrote them to get a better understanding myself :-)
What's missing are tests for multivariate `ConditionSet`s as @oscarbenjamin pointed out.

Reply:
Since they are arbitrary in name and have no significance for the computation of the expression and provide no help in knowing what to do to the expression, e.g. what to target for `subs`, they have no value. This is a YAGNI issue. Let's get rid of the routine added in this PR.

Reply:
>  Let's get rid of the routine added in this PR.

It's still expected by `Basic.as_dummy()`, no?
https://github.com/sympy/sympy/blob/32c28ebf80b6056b07c2a0d8bb4dc249741b86a4/sympy/core/basic.py#L587
So wouldn't removing it here entail rewriting `as_dummy` and also removing the now superfluous `bound_variables` also from `Sum`?

Reply:
OK, I see (again). The bound symbol routines let `as_dummy` know about conflicts and that is something that `expr.atoms(Symbol) - expr.free_symbols` cannot tell you. But you could do something like `expr.atoms(Symbol) & (expr.subs({i:Dummy() for i in expr.free_symbols})` to identify clashes...a dedicated `bound_symbols` seems to be a more targeted way of getting that information. So every object that has bound symbols should be responsible for defining them; I am not sure that it is necessary to do so recursively. So `ConditionSet(x, cond, base)` might only need to report `x`.

Reply:
I suppose that last one should have a `.as_dummy()` at the rhs too? Anyway, it's a good test ... that uncovers the following :-/
```
In [1]: C = ConditionSet((x, y), x <= y, S.Reals**2)

In [2]: C.as_dummy()
Out[2]: 
⎧                       2            ⎫
⎨(_1, _0) | (_1, _0) ∊ ℝ  ∧ (_1 ≤ _0)⎬
⎩                                    ⎭

In [3]: D = ConditionSet((y, x), y <= x, S.Reals**2)

In [4]: D.as_dummy()
Out[4]: 
⎧                       2            ⎫
⎨(_0, _1) | (_0, _1) ∊ ℝ  ∧ (_0 ≤ _1)⎬
⎩                                    ⎭
```
So, yes, one more issue. (and `.dummy_eq()` fails as expected)